pixl_core/src/core/upload.py

```diff
@@ -0,0 +1,119 @@
+#  Copyright (c) University College London Hospitals NHS Foundation Trust
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+"""Functionality to upload files to an endpoint."""
+
+from __future__ import annotations
+
+import ftplib
+import logging
+import os
+import ssl
+from datetime import datetime, timezone
+from ftplib import FTP_TLS
+from typing import TYPE_CHECKING, Any, BinaryIO
+
+if TYPE_CHECKING:
+    from pathlib import Path
+    from socket import socket
+
+from core._database import get_project_slug_from_db, update_exported_at_and_save
+
+logger = logging.getLogger(__name__)
+
+
+class ImplicitFtpTls(ftplib.FTP_TLS):
+    """
+    FTP_TLS subclass that automatically wraps sockets in SSL to support implicit FTPS.
+
+    https://stackoverflow.com/questions/12164470/python-ftp-implicit-tls-connection-issue
+    """
+
+    def __init__(self, *args: Any, **kwargs: Any) -> None:
+        """Create instance from parent class."""
+        super().__init__(*args, **kwargs)
+        self._sock: socket | None = None
+
+    @property
+    def sock(self) -> socket | None:
+        """Return the socket."""
+        return self._sock
+
+    @sock.setter
+    def sock(self, value: socket) -> None:
+        """When modifying the socket, ensure that it is ssl wrapped."""
+        if value is not None and not isinstance(value, ssl.SSLSocket):
+            value = self.context.wrap_socket(value)
+        self._sock = value
+
+
+def upload_dicom_image(local_file_path: Path, pseudo_anon_id: str) -> None:
+    """Top level way to upload an image."""
+    # Create the remote directory if it doesn't exist
+    # rename destination to {project-slug}/{study-pseduonymised-id}.zip
+    remote_directory = get_project_slug_from_db(pseudo_anon_id)
+
+    # Store the file using a binary handler
+    with local_file_path.open("rb") as file_content:
+        output_path = upload_content(
+            # wrong directory name, can get that from the image at least
+            file_content,
+            remote_dir=remote_directory,
+            remote_file=f"{pseudo_anon_id}.zip",
+        )
+
+    update_exported_at_and_save(pseudo_anon_id, datetime.now(tz=timezone.utc))
+    return output_path
+
+
+def upload_content(content: BinaryIO, *, remote_dir: str, remote_file: str) -> str:
```

Comment:
probably worth making this a private method so that we know we shouldn't be calling it directly